libobs: Fix duplicate audio bugs

pkviet opened the pull request:
### Description
This fixes two bugs:
1. increased audio level when the monitoring device is also captured with Desktop Audio.
2. increased audio level when a source is duplicated in a scene or through a nested scene.

### Motivation and Context
Bug 1 occurs in a standard setup for users:
- capture game audio through Desktop Audio,
- add a background audio which is monitored through the same device.
The background audio is then duplicated because it is captured through Desktop Audio.
This means that if one turns on and off the monitoring of a source, the audio level for any output will have an intrinsic fluctuation. (about +3 dBFS for music or white noise;  up to +6 dBFS for a low frequency tone)
This constrains the user to either get a device with enough separate outputs or to do complex routing with virtual cables;
or on windows to use the dedicated Application Audio Capture (which is still in beta).

Suppose for instance that one monitors in turn several audio sources (for instance speakers), their audio level would fluctuate from one to another. This makes the basic operation of monitoring audio a flawed process because it interferes with the audio which is output to a recording or a stream; the monitoring itself won't detect the feedback loop.

According to our UI chief (@Warchamp7 ), this has been one of the major blocks preventing his approval of a monitoring button in the audio meters.

Bug 2 was actually fixed in commits 8a38e33 and 72924ac. But it was fixed at the scene level and this left an edge case where a nested scene has a Show/Hide transition. It implies an increase of +6 dBFS for any duplicated source.
An example of valid use case is a Media Source or Browser source which is duplicated and cropped to different areas. Or a nested scene. We want to be able to transition from  scene A to scene B with nested scene A, without any change in audio level for instance, if the same audio is used across scenes. 

[1] [libobs: Mix audio of each source in a scene only once](https://github.com/obsproject/obs-studio/commit/8a38e3375b63692b01e5088d37f531f36a5d39f9)
[2] [libobs: Deduplicate audio for nested scenes/groups if not transitioning](https://github.com/obsproject/obs-studio/commit/72924ac1f3c4c55de038fe1f03a0e197a155d5e7)

We fix Bugs 1 and 2 in the core audio pipeline (in the audio_callback of obs-audio.c).

We fix Bug 1 in this manner:
- on the UI side, we detect whenever the same device is used for monitoring and Desktop Audio (or Desktop Audio 2).
- if it is so, core audio receives a signal to silence the audio output buffers of all sources which have monitoring_type MONITOR_AND_OUPUT.
- if the Desktop Audio is muted, the mute signal tells core audio to not silence monitored sources.

We fix Bug 2 in this manner:
- at each audio tick, when the audio tree is built, we tag duplicate audio sources;
- they are then bypassed in the audio rendering of any scenes or transitions they're involved;
- they're mixed in the final mix as root_nodes exactly like global audio sources.
This intrinsically prevents any duplication within a scene, across scenes and in transitions.
This is a better solution than the previous one because it is not limited to intra-scene deduplication.
This fixes for instance the edge case of Show/Hide transition to a nested scene.

### How Has This Been Tested?
This has been tested extensively on windows 11 pro 23H2 using:
- a sine tone at 1000 Hz, generated by lavfi filter in Media Source, of amplitude 1/8, which means -21.1 dBFS.
- the music from a splatoon clip.
The audio level of the sine was checked using the FFmpeg volumedetect filter:
ffmpeg -i '.\input.mp4' -filter:a volumedetect -f null /dev/null
Recordings were made in mp4 container with 16 bit pcm raw audio.
The waeveforms were checked and compared between OBS 31.0.3 (current release) and this PR in Reaper DAW.

The following tests were performed for Bug 1 (Desktop Audio duplication of monitored sources):

1. Scene with monitored sine source + Desktop Audio (monitoring device set at 100% level on the pc):
- left: obs current release: level is -15.4 dBFS or +5.7 dBFS above the expected level.
- right: this PR: level stays at nominal -21.1 dBFS

![dup1](https://github.com/user-attachments/assets/658b5b41-1a4c-4b75-bda6-21225b4740ce)

2. same scene with mute/unmute on the Desktop Audio:
- top: obs current release, one sees the +5.7 dBFS level variation.
- bottom: this PR. The audio level is undisturbed.
![dup2](https://github.com/user-attachments/assets/caa84d59-fc2a-41a5-a2b6-d343308277d2)

3. same scene with monitoring on and off on sine source:
- top: obs current release, one sees the +5.7 dBFS level variation.
- bottom: this PR. The audio level is undisturbed.
![dup3](https://github.com/user-attachments/assets/e2bb2b8a-4a40-492b-86ff-6b195fda0ba7)

4. scene with a Media Source clip (splatoon) of 5 seconds looped 4 times, monitored + Desktop Audio at 100% level on pc:
- on 2nd and 3rd loop, the Desktop Audio is muted;
- top: obs current release: level is higher on 1st and 4th loops.
- bottom: this PR: level is the same for all 4 loops (identical waveforms).
![dup4](https://github.com/user-attachments/assets/68d5e580-6414-476a-9eea-54454c5b822f)

For Bug 2, the following tests were done:
1. scene with two sines,
2. show/hide transition for a sine in a scene with duplicated sine, 
3. fade transition between a scene A with single sine to scene B w/ duplicated sine,
4. scene A with sine, nested in scene B which also has the sine as a source.
5. same as 4 with a show/hide transition on the sine in scene B (not shown, same result as 4.)
![dup5](https://github.com/user-attachments/assets/31562274-977f-463c-855b-76e50aea9a16)

6. same tests on a audio clip (splatoon): identical results. There is no regression on this PR.

7. same as 4. w/ a show/hide transition on the nested scene A:
- top: current release: there is a bump at the transition; level does not stay nominal;
- bottom: this PR: no bump, level stays nominal.
![dup6](https://github.com/user-attachments/assets/943134ae-f98a-49b2-8e27-ef168d12a098)

### Types of changes
- Bug fix (non-breaking change which fixes an issue)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Warchamp7 reviewed (approved):
I have been working with pkv on these changes so they have already gotten a good round of testing from myself on Windows 10.

These are deep audio pipeline changes though, so this is something we'll want a fair amount of testing on.

exeldro commented on libobs/obs.h:
```diff
@@ -449,6 +449,16 @@ EXPORT bool obs_get_audio_info(struct obs_audio_info *oai);
+/**
+ * Sets a bool warning that monitoring device coincides with one of DesktopAudio devices.
+ */
+EXPORT void obs_set_prevent_monitoring_duplication(int enable);
```
the comment above suggest a bool, but there only is an int and it is not added to the documentation

pkviet replied:
/**
 * Passes the channel index for a Desktop Audio (1 or 2) when its device coincides w/ the monitoring device.
 * Passes 0 if Destop Audio device and monitoring device are not the same.
 * Sets a bool warning core audio that monitoring device coincides with one of DesktopAudio devices.
 */

CyBeRoni commented:
> on the UI side, we detect whenever the same device is used for monitoring and Desktop Audio (or Desktop Audio 2).

Does this imply this fix only works for the two global Desktop Audio sources, and not if I add an Audio Output Capture source to a scene myself, along with monitored sources?

safocl commented on frontend/widgets/OBSBasic.cpp:
```diff
@@ -906,6 +906,56 @@ static inline void LogEncoders()
+bool GetDesktopAudioMute(int index)
+{
+	OBSSourceAutoRelease source = obs_get_output_source(index);
+	if (source)
```
I can't understand at all - why are their own implementations of unique_ptr used here until now? Are there any real objective reasons?

safocl replied:
https://github.com/obsproject/obs-studio/pull/12180

pkviet replied:
this is just checking for a nullptr ...
obs_get_output_source is ref counted hence the use of OBSSourceAutoRelease .
You're making wrong assumptions about the code base; there is not a single ownership of a source pointer in libobs; so your PR is also wrong.

safocl replied:
> this is just checking for a nullptr ... obs_get_output_source is ref counted hence the use of OBSSourceAutoRelease . You're making wrong assumptions about the code base; there is not a single ownership of a source pointer in libobs; so your PR is also wrong.

how exactly does this check for nullptr?

safocl replied:
https://godbolt.org/z/6Y7vhGWvb -- this example shows that this is not the case

safocl replied:
do you think that it is first cast via the type operator to a pointer?

safocl replied:
To do this you need to explicitly cast to the pointer:
OBSSourceAutoRelease source = obs_get_output_source(index);
if (static_cast<decltype(obs_get_output_source(index))>(source)) {}

safocl replied:
and if the pointer is `nullptr`, it will create UB if the release function dereferences the pointer:
https://github.com/obsproject/obs-studio/blob/b46e5bb1c08fb385d19f024199557c792a4e6126/libobs/obs.hpp#L77
destructor does not check for nullptr

safocl replied:
> To do this you need to explicitly cast to the pointer:
> 
> ```c++
> OBSSourceAutoRelease source = obs_get_output_source(index);
> if (static_cast<decltype(obs_get_output_source(index))>(source)) {}
> ```

although it does... hmm... I'm a little lost.

safocl replied:
> no you're wrong; just compile obs, and breakpoint, instead of making assumptions; the crash in your godbolt code is due to destructor of a nullptr. If you want to see the false branch of the nullptr taken, just add << std::flush; to all your std::cout lines ... Thee check of the nullptr works perfectly.

yes -- it's a kill and it doesn't give a normally working code

safocl commented on frontend/widgets/OBSBasic.cpp:
```diff
@@ -906,6 +906,56 @@ static inline void LogEncoders()
+QString GetDesktopAudioId(int index)
+{
+	OBSSourceAutoRelease source = obs_get_output_source(index);
+	const char *devId = nullptr;
+	OBSDataAutoRelease settings = nullptr;
+
+	if (source) {
+		settings = obs_source_get_settings(source);
+		if (settings)
+			devId = obs_data_get_string(settings, "device_id");
+	}
+
+	return QString(devId);
```
maybe:
```cpp
	QString devId;
	OBSDataAutoRelease settings = nullptr;

	if (source) {
		settings = obs_source_get_settings(source);
		if (settings)
			devId = obs_data_get_string(settings, "device_id");
	}

	return devId;
```
?

safocl commented on frontend/widgets/OBSBasic.cpp:
```diff
@@ -906,6 +906,56 @@ static inline void LogEncoders()
+
+int OBSBasic::MonDeviceIsDesktopAudioDevice()
+{
+	QString monDevId = config_get_string(Config(), "Audio", "MonitoringDeviceId");
+	QString desktopAudioDevice1 = GetDesktopAudioId(1);
+	QString desktopAudioDevice2 = GetDesktopAudioId(2);
```
maybe:
```cpp
	const QString monDevId = config_get_string(Config(), "Audio", "MonitoringDeviceId");
	const QString desktopAudioDevice1 = GetDesktopAudioId(1);
	const QString desktopAudioDevice2 = GetDesktopAudioId(2);
```
?

pkviet commented:
**Update**
Updated the PR to support monitoring deduplication with both global 'Desktop Audio' sources and 'Audio_Output_Capture' sources (wasapi and pulse).
Removed all UI code in favour of fully processing deduplication of monitored sources in core audio_callback.
The deduplication code for sources duplicated in scenes or across scenes is unchanged.

pkviet commented:
**Note**
The earlier version of this PR relied on UI to detect when Monitoring and Desktop Audio use the same device.
It might be still be a viable alternative to the latest version (although it doesn't cover Audio Output Capture).
It is available in this branch of my fork: https://github.com/pkviet/obs-studio/tree/audio-fixes-back
I can revert the PR to that earlier version if on review it is decided to be a better option.

PatTheMav reviewed (commented):
There's a whole lot of stuff happening in frontend code that I feel like it has no place there - the UI should be able to happily "do its thing" and it should be the monitoring system and/or the output system that are synching on this and prevent the duplication from happening.

I'll get into more detail on this later, but just to make sure I understand the expected outcome right:

If there is _any_ source capturing the output of "Device A" and the same "Device A" is then added as a monitoring device, then the _monitoring output_ for all of these capture sources is not supposed to be routed to the monitoring output (their contribution to the monitoring output is effectively "skipped")?

Wouldn't it be sufficient if the monitoring source during set up checks the device ID associated with the source it is being attached to and become "inert" when it is the same ID (and repeat this check for any update to the source it's attached to, so when the captured device is changed, the monitoring source will notice and re-enable monitoring output again)?

pkviet commented:
> There's a whole lot of stuff happening in frontend code that I feel like it has no place there - the UI should be able to happily "do its thing" and it should be the monitoring system and/or the output system that are synching on this and prevent the duplication from happening.
> 
> I'll get into more detail on this later, but just to make sure I understand the expected outcome right:
> 
> If there is _any_ source capturing the output of "Device A" and the same "Device A" is then added as a monitoring device, then the _monitoring output_ for all of these capture sources is not supposed to be routed to the monitoring output (their contribution to the monitoring output is effectively "skipped")?
> 
correct

> Wouldn't it be sufficient if the monitoring source during set up checks the device ID associated with the source it is being attached to and become "inert" when it is the same ID (and repeat this check for any update to the source it's attached to, so when the captured device is changed, the monitoring source will notice and re-enable monitoring output again)?

yes, this is what is done in the PR. Since the monitoring device changes are UI driven, the PR tracks that in UI. It is possible to do that purely in libobs, since the monitoring_device_id is stored in obs_core::audio struct; but we would need to check for all audio sources in the audio tree whether they are AOC (audio output capture sources) and in case they are, whether their device coincides with the monitoring device. This must be done at each audio tick.
As noted in the June 8 update to this PR, the pure libobs version of this PR does just that.
The reason why I opted for the current UI version are in summary:
- to make minimal changes to the core audio pipeline,
- to limit the friction by avoiding a function which checks at each audio tick every audio source in the audio tree and verifies they're not an AOC with a device identical with the monitoring device. In practice from audio tick to audio tick, it will be very rare that a new AOC coincides with the mon device. So it feels natural to make such checks on the UI side.
Nevertheless, it could be argued that such decisions belong to the backend (libobs) and not the UI, which should not have to know anything about such subtleties. So I am fine with switching to the pure libobs version if it so desired.

pkviet commented:
**Update**
Following comments, I have reverted to the pure libobs version. It is rebased on the other PR https://github.com/obsproject/obs-studio/pull/12268 ; it could be made to be completely self consistent but there would then be a rebase required of either 12268 or 12175 .

PatTheMav commented on libobs/obs-audio.c:
```diff
@@ -40,6 +40,45 @@ static void push_audio_tree(obs_source_t *parent, obs_source_t *source, void *p)
+	struct obs_core_audio *audio = p;
+	if (!audio->monitoring_device_name)
+		return;
+	if (strcmp(s->info.id, "wasapi_output_capture") == 0 || strcmp(s->info.id, "pulse_output_capture") == 0) {
```
Note that this also doesn't cover "Desktop Capture" via SCK, which will always capture all system audio _regardless of device_.

So if I play music to device A in an app, SCK will capture it. And if OBS' monitoring also outputs to device A you end up with duplicate audio again.

That situation is probably not automagically fixable from OBS' side, as OBS cannot know if any application produces "system audio" that is actually put out via OBS' monitoring device.

Similarly if I use application/window capture just for a single application, SCK will provide it to OBS regardless of the selected output device, so in that case using the same monitoring device will lead to duplication.

This might also affect Windows audio capture, but I haven't checked that one.

Just something to be aware of, `coreaudio_output_capture` would be the only source that could be automatically covered for via this PR.

pkviet commented on libobs/obs-audio.c:
```diff
@@ -509,11 +585,14 @@ bool audio_callback(void *param, uint64_t start_ts_in, uint64_t end_ts_in, uint6
+			// add top-level sources to audio tree
+			push_audio_tree(NULL, source, audio);
+			// check whether the source is an AOC & coincides with monitoring device
+			check_audio_output_source_is_monitoring_device(source, audio);
```
fixed

pkviet commented on libobs/obs-audio.c:
```diff
@@ -465,6 +506,41 @@ static inline void execute_audio_tasks(void)
+ */
+static inline bool should_silence_monitored_source(obs_source_t *source, struct obs_core_audio *audio)
+{
+	if (!audio->monitoring_duplicating_source)
+		return false;
+	bool output_capture_unmuted = !audio->monitoring_duplicating_source->user_muted;
```
done

pkviet commented on libobs/obs-audio.c:
```diff
@@ -465,6 +506,41 @@ static inline void execute_audio_tasks(void)
 	}
 }
 
+/* In case monitoring and an AOC (Audio Output Capture) source have the same device, one silences all the monitored
+ * sources unless the AOC is muted.
+ * The syncing between the mute state of the AOC source set in UI and libobs is a bit tricky. There is an intrinsic 
+ * positive delay of the AOC source w/ respect to monitored sources due to the audio OS processing (wasapi or pulse
+ * audio). Unfortunately the delay is machine dependent and we can only mitigate its effects. During tests, 
+ * src->user_muted worked better than src->muted, maybe because it is set earlier than the muted flag which allows to 
+ * keep a better sync ? W/ src->muted, unmuting the AOC led to a systematic level increase during one tick on my dev 
+ * machine for a sine tone (about +3 dBFS). In general we don't expect much difference though between user_muted and
+ * muted.
+ */
```
done

pkviet commented on libobs/obs-audio.c:
```diff
@@ -40,6 +40,47 @@ static void push_audio_tree(obs_source_t *parent, obs_source_t *source, void *p)
+
+/*
+ * This version of push_audio_tree checks whether any source is an Audio Output Capture source ('Desktop Audio',
+ * 'wasapi_output_capture' on windows, 'pulse_output_capture' on linux, 'coreaudio_output_capture' on macOS), & if the
```
done

RytoEX commented:
#12268 has been merged.

pkviet commented:
> #12268 has been merged.

rebased to master

pkviet commented on libobs/obs-audio.c:
```diff
@@ -500,6 +526,42 @@ static inline void execute_audio_tasks(void)
+ * OS processing (wasapi, pulseaudio or coreaudio). Unfortunately the delay is machine dependent and we can only 
+ * mitigate its effects. During tests, src->user_muted worked better than src->muted, maybe because it is set earlier 
+ * than the muted flag which allows to keep a better sync ? With src->muted, unmuting the 'Audio Output Capture' led to
+ * a systematic level increase during one tick on my (pkv's) dev machine for a sine tone (about +3 dBFS). In general we
```
changed to: the dev machine

pkviet replied:
changed to your suggestion

pkviet commented on libobs/obs-audio.c:
```diff
@@ -543,13 +607,19 @@ bool audio_callback(void *param, uint64_t start_ts_in, uint64_t end_ts_in, uint6
 			/* add top - level sources to audio tree */
 			push_audio_tree(NULL, source, audio);
+
+			/* Check whether the source is an AOC & coincides with monitoring device */
```
done

RytoEX commented:
> note: no matter how many rebases on master i do, GH keeps saying : This branch is out-of-date with the base branch ...

If I recall correctly, it's a mistake on the GitHub UI. It will resolve once the PR is approved.

pkviet commented on libobs/obs-audio.c:
```diff
@@ -500,6 +526,42 @@ static inline void execute_audio_tasks(void)
+ * is an intrinsic positive delay of the 'Audio Output Capture' source w/ respect to monitored sources due to the audio
+ * OS processing (wasapi, pulseaudio or coreaudio). Unfortunately the delay is machine dependent and we can only 
+ * mitigate its effects. During tests, src->user_muted worked better than src->muted, maybe because it is set earlier 
+ * than the muted flag which allows to keep a better sync ? With src->muted, unmuting the 'Audio Output Capture' led to
```
it was to mark that this was a temptative guess, but that i'm not sure

RytoEX replied:
In general, we might use "(?)" for that, or omit it entirely.

pkviet replied:
Oh ok. Removing the aiestion mark then.

pkviet commented:
most recent comments addressed and PR force pushed.
